### PR TITLE
[FIRRTL] Add list concatenation operation.

### PR DESCRIPTION
**Pull request by mikeurbach**
This operation produces a list by concatenating lists of the same
type. The operation definition and a simple round trip test have been
added. This is used to compose lists, which is a key feature to enable
hierarchical composition of Properties.

### CODE REVIEW

**Review by prithayan — APPROVED** [review]

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td [hunk 1]
```diff
@@ -1154,6 +1154,23 @@ def ListCreateOp : FIRRTLOp<"list.create", [Pure, SameTypeOperands]> {
   let hasVerifier = 1;
 }
 
+def ListConcatOp : FIRRTLOp<"list.concat", [Pure, SameOperandsAndResultType]> {
```
Comment (dtzSiFive) [inline include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:1157]: Add to FIRRTLVisitor.h please.

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td [hunk 2]
```diff
@@ -1154,6 +1154,23 @@ def ListCreateOp : FIRRTLOp<"list.create", [Pure, SameTypeOperands]> {
...
+    Example:
+    ```mlir
+    %3 = firrtl.list.concat %0, %1, %2 : !firrtl.list<string>
+    ```
+  }];
+
+  let arguments = (ins Variadic<ListType>:$subLists);
+  let results = (outs ListType:$result);
+
+  let assemblyFormat = "$subLists attr-dict `:` type($result)";
```
Comment (dtzSiFive) [inline include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:1171]: Might be neat to give this a folder in the future for IR simplification (and use w/IMCP) but that can come later.

**Review by dtzSiFive — APPROVED** [review]

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td (hunk 1 above)
Comment (mikeurbach) [reply include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:1157]: Oops, yep will do, thanks!

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td (hunk 2 above)
Comment (mikeurbach) [reply include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:1171]: Could be interesting! I think there are lot of interesting optimizations we could explore on properties.

Comment (dtzSiFive) [reply include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:1171]: Yeah just thinking about cleaning up the trivial stuff that literally constant-folds.  I'm not sure we're likely to feed this constants at first/soon and yep lots of interesting things for sure! Look forward to it, LMK if want some assistance!

Looks like integer ops have issues filed for adding folders there, but quick check on some internal usage suggests these wouldn't help much just yet anyway.

Anyway stray thought, GLHF!

**Review by leonardt — APPROVED** [review]

**Pull request merged**